### PR TITLE
KnotMock implementation

### PR DESCRIPTION
**Pull request by Skejven**
Mocks for Knot.x Knots :)
---

I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/MockKnotHandler.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  private static final String RESPONSE_KEY = "clientResponse";
+  private static final JsonObject ERROR_RESPONSE = new JsonObject().put(RESPONSE_KEY, new JsonObject().put("statusCode", 500));
+  private static final JsonObject NOT_FOUND = new JsonObject().put(RESPONSE_KEY, new JsonObject().put("statusCode", 404));
+  private static final String MESSAGE_REQUEST_KEY = "request";
+  private static final List<String> MOCK_RESPONSE_KEYS = Arrays.asList("clientResponse", "fragments", "transition");
+  private static final Map<String, JsonObject> DEFAULT_RESPONSES = ImmutableMap.of(
+      "clientResponse", new JsonObject().put("statusCode", 200),
+      "fragments", new JsonObject(),
+      "transition", new JsonObject()
+  );
```
Comment (marcinczeczko) [inline knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/MockKnotHandler.java]: Transition should be just a string

Comment (Skejven) [inline knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/MockKnotHandler.java]: Ok, thanks, I will fix that

File: knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  RESPONSE("clientResponse") {
+    @Override
+    Object defaultValue() {
+      return new JsonObject().put("statusCode", 200);
+    }
+  },
+  REQUEST("clientRequest") {
+    @Override
+    Object defaultValue() {
+      return null;
```
Comment (marcinczeczko) [inline knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java]: maybe just return empty JsonObject()

File: knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  FRAGMENTS("fragments") {
+    @Override
+    Object defaultValue() {
+      return new JsonArray();
+    }
+  },
+  TRANSITION("transition") {
+    @Override
+    Object defaultValue() {
+      return null;
```
Comment (marcinczeczko) [inline knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java]: In my server routing I'm using Optional to carry on transitions. so maybe simply return Optional.empty() here

File: knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  Observable<Object> mockValue(FileSystem fileSystem, String resourcePath) {
+    return fileSystem.openObservable(resourcePath, new OpenOptions().setCreate(false).setWrite(false))
+        .flatMap(this::processFile)
+        .map(this::toJson);
+  }
+
+  abstract Object defaultValue();
+
+  private Object toJson(Buffer buffer) {
+    return buffer.toString().charAt(0) == '{' ? buffer.toJsonObject() : buffer.toJsonArray();
```
Comment (marcinczeczko) [inline knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java]: toString().trim() - to assure no white space as first character

File: knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java (hunk 2 above)
Comment (Skejven) [inline knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java]: Null was just to indicate that there is no default value. I will switch to `Optional`.

File: knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java (hunk 3 above)
Comment (Skejven) [inline knotx-example/knotx-mocks/src/main/java/com/cognifide/knotx/mocks/knot/KnotContextKeys.java]: Changed.

**Pull request merged**